File: 0150-evaluate-reverse-polish-notation/0150-evaluate-reverse-polish-notation.py
from operator import add
from operator import sub
from operator import mul


class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        operators = {
            "+": add,
            "-": sub,
            "*": mul,
            "/": lambda x, y: int(x / y)
        }
        stack = []
        for token in tokens:
            if token not in operators:
                stack.append(int(token))
                continue
            n2 = stack.pop()
            n1 = stack.pop()
            stack.append(operators[token](n1, n2))
        return stack[0]

File: 0150-evaluate-reverse-polish-notation/0150-evaluate-reverse-polish-notation.cpp
class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        static unordered_map<string, function<int(int, int)>> op{{
            {"+", plus<int>()},
            {"-", minus<int>()},
            {"*", multiplies<int>()},
            {"/", divides<int>()},
        }};

        stack<int> stk;

        for (const auto& token : tokens) {
            if (isdigit(token.back())) {
                stk.push(stoi(token));
            } else {
                int rhs = stk.top();
                stk.pop();
                int lhs = stk.top();
                stk.pop();
                stk.push(op[token](lhs, rhs));
            }
        }
        return stk.top();
    }
};
